From 0be3f48f76c38da7eb04a3afe387ef3cec29af2d Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Wed, 26 Jun 2024 20:24:30 +0200
Subject: [PATCH] Implement Dogecoin's DAA (#11)

Add `Consensus::DaaParams`, because the DAA parameters on Dogecoin vary based on height.

Dogecoin solves this adding a height parameter to `GetConsensus`, but this seems like an ugly solution, instead we add `DaaParamsAtHeight` to `Consensus::Params`.

(cherry picked from commit a122ebc5c0d8408fecbaf19f817cb14631a5f434)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index d9a3430ae3a22eea297e91cf8c34906cb5186d78..5f3f37f9d06316973fab6134da533ba21c967a53 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -587,6 +587,7 @@ add_library(bitcoinconsensus
 	consensus/activation.cpp
 	consensus/amount.cpp
 	consensus/consensus.cpp
+	consensus/params.cpp
 	consensus/tx_check.cpp
 )
 
@@ -690,7 +691,6 @@ add_library(server
 	policy/packages.cpp
 	policy/settings.cpp
 	pow/auxpow.cpp
-	pow/eda.cpp
 	pow/pow.cpp
 	rest.cpp
 	rpc/abc.cpp
@@ -854,6 +854,7 @@ if(BUILD_CHAINSTATE)
 		consensus/amount.cpp
 		consensus/consensus.cpp
 		consensus/merkle.cpp
+		consensus/params.cpp
 		consensus/tx_check.cpp
 		consensus/tx_verify.cpp
 		core_read.cpp
@@ -877,7 +878,6 @@ if(BUILD_CHAINSTATE)
 		policy/policy.cpp
 		policy/settings.cpp
 		pow/auxpow.cpp
-		pow/eda.cpp
 		pow/pow.cpp
 		primitives/block.cpp
 		primitives/transaction.cpp
diff --git a/src/consensus/params.cpp b/src/consensus/params.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..eacd84d13e18fa2dae3b77d26c8b755830767450
--- /dev/null
+++ b/src/consensus/params.cpp
@@ -0,0 +1,45 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <consensus/activation.h>
+
+#include <consensus/params.h>
+
+namespace Consensus {
+
+DaaParams Params::DaaParamsAtHeight(int32_t nHeight) const {
+    DaaParams daaParams;
+
+    const bool hasDigishield = IsDigishieldEnabled(*this, nHeight);
+
+    daaParams.fPowAllowMinDifficultyBlocks = enableTestnetMinDifficulty;
+    // Blocks 145000 - 157499 have fPowAllowMinDifficultyBlocks disabled
+    if (enableTestnetMinDifficulty && hasDigishield && nHeight < 157500) {
+        daaParams.fPowAllowMinDifficultyBlocks = false;
+    }
+
+    daaParams.fDigishieldDifficultyCalculation = hasDigishield;
+
+    daaParams.nPowTargetTimespan = hasDigishield ? 60 : 4 * 60 * 60;
+
+    if (hasDigishield) {
+        daaParams.nMinTimespan =
+            daaParams.nPowTargetTimespan - (daaParams.nPowTargetTimespan / 4);
+        daaParams.nMaxTimespan =
+            daaParams.nPowTargetTimespan + (daaParams.nPowTargetTimespan / 2);
+    } else if (nHeight > 10000) {
+        daaParams.nMinTimespan = daaParams.nPowTargetTimespan / 4;
+        daaParams.nMaxTimespan = daaParams.nPowTargetTimespan * 4;
+    } else if (nHeight > 5000) {
+        daaParams.nMinTimespan = daaParams.nPowTargetTimespan / 8;
+        daaParams.nMaxTimespan = daaParams.nPowTargetTimespan * 4;
+    } else {
+        daaParams.nMinTimespan = daaParams.nPowTargetTimespan / 16;
+        daaParams.nMaxTimespan = daaParams.nPowTargetTimespan * 4;
+    }
+
+    return daaParams;
+}
+
+} // namespace Consensus
diff --git a/src/consensus/params.h b/src/consensus/params.h
index 288bed6c9b19c41a6e8e86728a9ca7864f81dc37..422a37cdbb5b0e9cc8213d081ddbaa2d7347f012 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -28,6 +28,14 @@ constexpr bool ValidDeployment(BuriedDeployment dep) {
     return dep <= DEPLOYMENT_CSV;
 }
 
+struct DaaParams {
+    bool fPowAllowMinDifficultyBlocks;
+    bool fDigishieldDifficultyCalculation;
+    int64_t nPowTargetTimespan;
+    int64_t nMinTimespan;
+    int64_t nMaxTimespan;
+};
+
 /**
  * Parameters that influence chain consensus.
  */
@@ -75,6 +83,9 @@ struct Params {
     /** Coinbase maturity before Digishield */
     int32_t initialCoinbaseMaturity;
 
+    /** Testnet allows min difficulty for some block height ranges */
+    bool enableTestnetMinDifficulty;
+
     /** Enable or disable the miner fund by default */
     bool enableMinerFund;
 
@@ -83,16 +94,14 @@ struct Params {
 
     /** Proof of work parameters */
     uint256 powLimit;
-    bool fPowAllowMinDifficultyBlocks;
     bool fPowNoRetargeting;
     int64_t nDAAHalfLife;
     int64_t nPowTargetSpacing;
-    int64_t nPowTargetTimespan;
     std::chrono::seconds PowTargetSpacing() const {
         return std::chrono::seconds{nPowTargetSpacing};
     }
-    int64_t DifficultyAdjustmentInterval() const {
-        return nPowTargetTimespan / nPowTargetSpacing;
+    int64_t DifficultyAdjustmentInterval(const DaaParams &daaParams) const {
+        return daaParams.nPowTargetTimespan / nPowTargetSpacing;
     }
     uint256 nMinimumChainWork;
     BlockHash defaultAssumeValid;
@@ -112,6 +121,8 @@ struct Params {
         } // no default case, so the compiler can warn about missing cases
         return std::numeric_limits<int>::max();
     }
+
+    DaaParams DaaParamsAtHeight(int32_t nHeight) const;
 };
 
 } // namespace Consensus
diff --git a/src/kernel/chainparams.cpp b/src/kernel/chainparams.cpp
index 032ce2fb71981686226eec88ce53b037346b6f85..c3a07febab82d0162da8953e5affd84e54627c12 100644
--- a/src/kernel/chainparams.cpp
+++ b/src/kernel/chainparams.cpp
@@ -93,14 +93,15 @@ public:
         consensus.powLimit = uint256S("0x00000fffffffffffffffffffffffffffffffff"
                                       "ffffffffffffffffffffffffff");
         // two weeks
-        consensus.nPowTargetTimespan = 14 * 24 * 60 * 60;
-        consensus.nPowTargetSpacing = 10 * 60;
-        consensus.fPowAllowMinDifficultyBlocks = false;
+        consensus.nPowTargetSpacing = 60; // one minute
         consensus.fPowNoRetargeting = false;
 
         // two days
         consensus.nDAAHalfLife = 2 * 24 * 60 * 60;
 
+        // Disable min difficulty rules on mainnet
+        consensus.enableTestnetMinDifficulty = false;
+
         // The miner fund is disabled by default on Dogecoin mainnet.
         consensus.enableMinerFund = false;
 
@@ -247,14 +248,15 @@ public:
         consensus.powLimit = uint256S("0x00000fffffffffffffffffffffffffffffffff"
                                       "ffffffffffffffffffffffffff");
         // two weeks
-        consensus.nPowTargetTimespan = 14 * 24 * 60 * 60;
         consensus.nPowTargetSpacing = 10 * 60;
-        consensus.fPowAllowMinDifficultyBlocks = true;
         consensus.fPowNoRetargeting = false;
 
         // two days
         consensus.nDAAHalfLife = 2 * 24 * 60 * 60;
 
+        // Enable min difficulty rules on testnet
+        consensus.enableTestnetMinDifficulty = true;
+
         // The miner fund is disabled by default on testnet.
         consensus.enableMinerFund = false;
 
@@ -383,14 +385,15 @@ public:
         consensus.powLimit = uint256S(
             "7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
         // two weeks
-        consensus.nPowTargetTimespan = 14 * 24 * 60 * 60;
         consensus.nPowTargetSpacing = 10 * 60;
-        consensus.fPowAllowMinDifficultyBlocks = true;
         consensus.fPowNoRetargeting = true;
 
         // two days
         consensus.nDAAHalfLife = 2 * 24 * 60 * 60;
 
+        // Disable min difficulty rules on regtest
+        consensus.enableTestnetMinDifficulty = false;
+
         // The miner fund is disabled by default on regtest.
         consensus.enableMinerFund = false;
 
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 1813abc1041709accc1b6ef671beadcee3150412..8821e74cfc403d844329bc190a687f15dd35cd6c 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -89,8 +89,8 @@ static constexpr auto HEADERS_RESPONSE_TIME{2min};
 static constexpr int32_t MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT = 4;
 /** Timeout for (unprotected) outbound peers to sync to our chainwork */
 static constexpr auto CHAIN_SYNC_TIMEOUT{20min};
-/** How frequently to check for stale tips */
-static constexpr auto STALE_CHECK_INTERVAL{10min};
+/** How frequently to check for stale tips (Dogecoin: 1 minute block time) */
+static constexpr auto STALE_CHECK_INTERVAL{1min};
 /** How frequently to check for extra outbound peers and disconnect. */
 static constexpr auto EXTRA_PEER_CHECK_INTERVAL{45s};
 /**
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index 359123e7e5a11528171f94127528e48162fa226d..4b3257d48752189a1800587446bf5b4f20876563 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -46,7 +46,9 @@ int64_t UpdateTime(CBlockHeader *pblock, const CChainParams &chainParams,
     }
 
     // Updating time can change work required on testnet:
-    if (chainParams.GetConsensus().fPowAllowMinDifficultyBlocks) {
+    if (chainParams.GetConsensus()
+            .DaaParamsAtHeight(pindexPrev->nHeight + 1)
+            .fPowAllowMinDifficultyBlocks) {
         pblock->nBits = GetNextWorkRequired(pindexPrev, pblock, chainParams);
     }
 
diff --git a/src/pow/eda.cpp b/src/pow/eda.cpp
deleted file mode 100644
index 67657b3a23980e3feee9981c35c502312beda4ec..0000000000000000000000000000000000000000
--- a/src/pow/eda.cpp
+++ /dev/null
@@ -1,130 +0,0 @@
-// Copyright (c) 2017-2020 The Bitcoin developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-#include <pow/eda.h>
-
-#include <arith_uint256.h>
-#include <chain.h>
-#include <consensus/params.h>
-
-/**
- * Do difficulty adjustement Satoshi's way.
- */
-uint32_t CalculateNextWorkRequired(const CBlockIndex *pindexPrev,
-                                   int64_t nFirstBlockTime,
-                                   const Consensus::Params &params) {
-    // Limit adjustment step
-    int64_t nActualTimespan = pindexPrev->GetBlockTime() - nFirstBlockTime;
-    if (nActualTimespan < params.nPowTargetTimespan / 4) {
-        nActualTimespan = params.nPowTargetTimespan / 4;
-    }
-
-    if (nActualTimespan > params.nPowTargetTimespan * 4) {
-        nActualTimespan = params.nPowTargetTimespan * 4;
-    }
-
-    // Retarget
-    const arith_uint256 bnPowLimit = UintToArith256(params.powLimit);
-    arith_uint256 bnNew;
-    bnNew.SetCompact(pindexPrev->nBits);
-    bnNew *= nActualTimespan;
-    bnNew /= params.nPowTargetTimespan;
-
-    if (bnNew > bnPowLimit) {
-        bnNew = bnPowLimit;
-    }
-
-    return bnNew.GetCompact();
-}
-
-/**
- * Compute the next required proof of work using the legacy Bitcoin difficulty
- * adjustment + Emergency Difficulty Adjustment (EDA).
- */
-uint32_t GetNextEDAWorkRequired(const CBlockIndex *pindexPrev,
-                                const CBlockHeader *pblock,
-                                const Consensus::Params &params) {
-    // Only change once per difficulty adjustment interval
-    uint32_t nHeight = pindexPrev->nHeight + 1;
-    if (nHeight % params.DifficultyAdjustmentInterval() == 0) {
-        // Go back by what we want to be 14 days worth of blocks
-        assert(nHeight >= params.DifficultyAdjustmentInterval());
-        uint32_t nHeightFirst = nHeight - params.DifficultyAdjustmentInterval();
-        const CBlockIndex *pindexFirst = pindexPrev->GetAncestor(nHeightFirst);
-        assert(pindexFirst);
-
-        return CalculateNextWorkRequired(pindexPrev,
-                                         pindexFirst->GetBlockTime(), params);
-    }
-
-    const uint32_t nProofOfWorkLimit =
-        UintToArith256(params.powLimit).GetCompact();
-
-    if (params.fPowAllowMinDifficultyBlocks) {
-        // Special difficulty rule for testnet:
-        // If the new block's timestamp is more than 2* 10 minutes then allow
-        // mining of a min-difficulty block.
-        if (pblock->GetBlockTime() >
-            pindexPrev->GetBlockTime() + 2 * params.nPowTargetSpacing) {
-            return nProofOfWorkLimit;
-        }
-
-        // Return the last non-special-min-difficulty-rules-block
-        const CBlockIndex *pindex = pindexPrev;
-        while (pindex->pprev &&
-               pindex->nHeight % params.DifficultyAdjustmentInterval() != 0 &&
-               pindex->nBits == nProofOfWorkLimit) {
-            pindex = pindex->pprev;
-        }
-
-        return pindex->nBits;
-    }
-
-    return pindexPrev->nBits;
-}
-
-// Check that on difficulty adjustments, the new difficulty does not increase
-// or decrease beyond the permitted limits.
-bool PermittedEDADifficultyTransition(const Consensus::Params &params,
-                                      uint32_t old_nbits,
-                                      arith_uint256 new_target) {
-    int64_t smallest_timespan = params.nPowTargetTimespan / 4;
-    int64_t largest_timespan = params.nPowTargetTimespan * 4;
-
-    const arith_uint256 pow_limit = UintToArith256(params.powLimit);
-
-    // Calculate the largest difficulty value possible:
-    arith_uint256 largest_difficulty_target;
-    largest_difficulty_target.SetCompact(old_nbits);
-    largest_difficulty_target *= largest_timespan;
-    largest_difficulty_target /= params.nPowTargetTimespan;
-
-    if (largest_difficulty_target > pow_limit) {
-        largest_difficulty_target = pow_limit;
-    }
-
-    // Round and then compare this new calculated value to what is
-    // observed.
-    arith_uint256 maximum_new_target;
-    maximum_new_target.SetCompact(largest_difficulty_target.GetCompact());
-    if (maximum_new_target < new_target) {
-        return false;
-    }
-
-    // Calculate the smallest difficulty value possible:
-    arith_uint256 smallest_difficulty_target;
-    smallest_difficulty_target.SetCompact(old_nbits);
-    smallest_difficulty_target *= smallest_timespan;
-    smallest_difficulty_target /= params.nPowTargetTimespan;
-
-    if (smallest_difficulty_target > pow_limit) {
-        smallest_difficulty_target = pow_limit;
-    }
-
-    // Round and then compare this new calculated value to what is
-    // observed.
-    arith_uint256 minimum_new_target;
-    minimum_new_target.SetCompact(smallest_difficulty_target.GetCompact());
-    return new_target >= minimum_new_target;
-}
diff --git a/src/pow/eda.h b/src/pow/eda.h
deleted file mode 100644
index e556c704663ea88765516b37a8b682df1cbf9792..0000000000000000000000000000000000000000
--- a/src/pow/eda.h
+++ /dev/null
@@ -1,38 +0,0 @@
-// Copyright (c) 2017-2020 The Bitcoin developers
-// Distributed under the MIT software license, see the accompanying
-// file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
-#ifndef BITCOIN_POW_EDA_H
-#define BITCOIN_POW_EDA_H
-
-#include <cstdint>
-
-class arith_uint256;
-class CBlockHeader;
-class CBlockIndex;
-
-namespace Consensus {
-struct Params;
-}
-
-uint32_t CalculateNextWorkRequired(const CBlockIndex *pindexPrev,
-                                   int64_t nFirstBlockTime,
-                                   const Consensus::Params &params);
-
-uint32_t GetNextEDAWorkRequired(const CBlockIndex *pindexPrev,
-                                const CBlockHeader *pblock,
-                                const Consensus::Params &params);
-
-/**
- * Return false if the proof-of-work requirement specified by new_target is not
- * possible, given the proof-of-work on the prior block as specified by
- * old_nbits.
- *
- * This function only checks that the new value is within a factor of 4 of the
- * old value for blocks.
- */
-bool PermittedEDADifficultyTransition(const Consensus::Params &params,
-                                      uint32_t old_nbits,
-                                      arith_uint256 new_target);
-
-#endif // BITCOIN_POW_EDA_H
diff --git a/src/pow/pow.cpp b/src/pow/pow.cpp
index 13e341fe8a0bfd4e24ded75893606318a753b6fb..58ed1f5b250aa9cefd5895543609f8eba5bbfdb0 100644
--- a/src/pow/pow.cpp
+++ b/src/pow/pow.cpp
@@ -12,9 +12,24 @@
 #include <common/system.h>
 #include <consensus/activation.h>
 #include <consensus/params.h>
-#include <pow/eda.h>
 #include <primitives/blockhash.h>
 
+// Dogecoin: Normally minimum difficulty blocks can only occur in between
+// retarget blocks. However, once we introduce Digishield every block is
+// a retarget, so we need to handle minimum difficulty on all blocks.
+bool AllowDigishieldMinDifficultyForBlock(
+    const CBlockIndex *pindexLast, const CBlockHeader *pblock,
+    const Consensus::Params &params, const Consensus::DaaParams &daaParams) {
+    // check if the chain allows minimum difficulty blocks
+    if (!daaParams.fPowAllowMinDifficultyBlocks) {
+        return false;
+    }
+
+    // Allow for a minimum block time if the elapsed time > 2*nTargetSpacing
+    return (pblock->GetBlockTime() >
+            pindexLast->GetBlockTime() + params.nPowTargetSpacing * 2);
+}
+
 uint32_t GetNextWorkRequired(const CBlockIndex *pindexPrev,
                              const CBlockHeader *pblock,
                              const CChainParams &chainParams) {
@@ -28,7 +43,95 @@ uint32_t GetNextWorkRequired(const CBlockIndex *pindexPrev,
         return pindexPrev->nBits;
     }
 
-    return GetNextEDAWorkRequired(pindexPrev, pblock, params);
+    unsigned int nProofOfWorkLimit =
+        UintToArith256(params.powLimit).GetCompact();
+
+    const int32_t nHeight = pindexPrev->nHeight;
+    const Consensus::DaaParams daaParams = params.DaaParamsAtHeight(nHeight);
+
+    // Dogecoin: Special rules for minimum difficulty blocks with Digishield
+    if (AllowDigishieldMinDifficultyForBlock(pindexPrev, pblock, params,
+                                             daaParams)) {
+        // Special difficulty rule for testnet:
+        // If the new block's timestamp is more than 2* nTargetSpacing minutes
+        // then allow mining of a min-difficulty block.
+        return nProofOfWorkLimit;
+    }
+
+    // Only change once per difficulty adjustment interval
+    const int64_t defaultInterval =
+        params.DifficultyAdjustmentInterval(daaParams);
+    const int64_t difficultyAdjustmentInterval =
+        daaParams.fDigishieldDifficultyCalculation ? 1 : defaultInterval;
+    if ((nHeight + 1) % difficultyAdjustmentInterval != 0) {
+        if (daaParams.fPowAllowMinDifficultyBlocks) {
+            // Special difficulty rule for testnet:
+            // If the new block's timestamp is more than 2* 10 minutes
+            // then allow mining of a min-difficulty block.
+            if (pblock->GetBlockTime() >
+                pindexPrev->GetBlockTime() + params.nPowTargetSpacing * 2) {
+                return nProofOfWorkLimit;
+            } else {
+                // Return the last non-special-min-difficulty-rules-block
+                const CBlockIndex *pindex = pindexPrev;
+                while (pindex->pprev &&
+                       pindex->nHeight % defaultInterval != 0 &&
+                       pindex->nBits == nProofOfWorkLimit) {
+                    pindex = pindex->pprev;
+                }
+                return pindex->nBits;
+            }
+        }
+        return pindexPrev->nBits;
+    }
+
+    // Litecoin: This fixes an issue where a 51% attack can change difficulty at
+    // will. Go back the full period unless it's the first retarget after
+    // genesis.
+    int32_t blocksToGoBack = difficultyAdjustmentInterval - 1;
+    if (nHeight + 1 != difficultyAdjustmentInterval) {
+        blocksToGoBack = difficultyAdjustmentInterval;
+    }
+
+    // Go back by what we want to be 14 days worth of blocks
+    int32_t nHeightFirst = pindexPrev->nHeight - blocksToGoBack;
+    assert(nHeightFirst >= 0);
+
+    const CBlockIndex *pindexFirst = pindexPrev->GetAncestor(nHeightFirst);
+    assert(pindexFirst);
+
+    const int64_t retargetTimespan = daaParams.nPowTargetTimespan;
+    const int64_t nActualTimespan =
+        pindexPrev->GetBlockTime() - pindexFirst->GetBlockTime();
+    int64_t nModulatedTimespan = nActualTimespan;
+
+    if (daaParams.fDigishieldDifficultyCalculation) {
+        // DigiShield implementation
+        nModulatedTimespan =
+            retargetTimespan + (nModulatedTimespan - retargetTimespan) / 8;
+    }
+
+    // Limit adjustment step
+    if (nModulatedTimespan < daaParams.nMinTimespan) {
+        nModulatedTimespan = daaParams.nMinTimespan;
+    } else if (nModulatedTimespan > daaParams.nMaxTimespan) {
+        nModulatedTimespan = daaParams.nMaxTimespan;
+    }
+
+    // Retarget
+    const arith_uint256 bnPowLimit = UintToArith256(params.powLimit);
+    arith_uint256 bnNew;
+    arith_uint256 bnOld;
+    bnNew.SetCompact(pindexPrev->nBits);
+    bnOld = bnNew;
+    bnNew *= nModulatedTimespan;
+    bnNew /= retargetTimespan;
+
+    if (bnNew > bnPowLimit) {
+        bnNew = bnPowLimit;
+    }
+
+    return bnNew.GetCompact();
 }
 
 // Check that on difficulty adjustments, the new difficulty does not increase
@@ -36,7 +139,9 @@ uint32_t GetNextWorkRequired(const CBlockIndex *pindexPrev,
 bool PermittedDifficultyTransition(const Consensus::Params &params,
                                    int64_t height, uint32_t old_nbits,
                                    uint32_t new_nbits) {
-    if (params.fPowAllowMinDifficultyBlocks) {
+    const Consensus::DaaParams daaParams = params.DaaParamsAtHeight(height - 1);
+
+    if (daaParams.fPowAllowMinDifficultyBlocks || params.fPowNoRetargeting) {
         return true;
     }
 
@@ -44,16 +149,16 @@ bool PermittedDifficultyTransition(const Consensus::Params &params,
     // assuming the initial difficulty was valid, so bail out early.
     // The initial difficulty is valid because we start from the genesis block
     // and we stop calling this function as soon as it returns false.
-    // This avoids further computation for most blocks prior to the BCH fork.
+    // This avoids further computation for most blocks prior to DAA forks.
     if (old_nbits == new_nbits) {
         return true;
     }
 
-    // Prior to the UAHF, the difficulty could change only every 2016 blocks,
+    // Prior to Digishield, the difficulty could change only every 240 blocks,
     // so we can bail out early if we observe a difficulty change at an
     // unexpected block height.
-    if (!IsUAHFenabled(params, height - 1) &&
-        height % params.DifficultyAdjustmentInterval() != 0) {
+    if (!IsDigishieldEnabled(params, height - 1) &&
+        height % params.DifficultyAdjustmentInterval(daaParams) != 0) {
         return false;
     }
 
@@ -63,14 +168,47 @@ bool PermittedDifficultyTransition(const Consensus::Params &params,
         return false;
     }
 
-    // The newer difficulty adjustement algorithms (CW-144 and Aserti32d) do
-    // not have other hard rules for permitted difficulty transitions.
-    if (IsDAAEnabled(params, height - 1)) {
-        return true;
+    int64_t smallest_timespan = daaParams.nMinTimespan;
+    int64_t largest_timespan = daaParams.nMaxTimespan;
+
+    const arith_uint256 pow_limit = UintToArith256(params.powLimit);
+    observed_new_target.SetCompact(new_nbits);
+
+    // Calculate the largest difficulty value possible:
+    arith_uint256 largest_difficulty_target;
+    largest_difficulty_target.SetCompact(old_nbits);
+    largest_difficulty_target *= largest_timespan;
+    largest_difficulty_target /= daaParams.nPowTargetTimespan;
+
+    if (largest_difficulty_target > pow_limit) {
+        largest_difficulty_target = pow_limit;
     }
 
-    return PermittedEDADifficultyTransition(params, old_nbits,
-                                            observed_new_target);
+    // Round and then compare this new calculated value to what is observed.
+    arith_uint256 maximum_new_target;
+    maximum_new_target.SetCompact(largest_difficulty_target.GetCompact());
+    if (maximum_new_target < observed_new_target) {
+        return false;
+    }
+
+    // Calculate the smallest difficulty value possible:
+    arith_uint256 smallest_difficulty_target;
+    smallest_difficulty_target.SetCompact(old_nbits);
+    smallest_difficulty_target *= smallest_timespan;
+    smallest_difficulty_target /= daaParams.nPowTargetTimespan;
+
+    if (smallest_difficulty_target > pow_limit) {
+        smallest_difficulty_target = pow_limit;
+    }
+
+    // Round and then compare this new calculated value to what is observed.
+    arith_uint256 minimum_new_target;
+    minimum_new_target.SetCompact(smallest_difficulty_target.GetCompact());
+    if (minimum_new_target > observed_new_target) {
+        return false;
+    }
+
+    return true;
 }
 
 bool CheckProofOfWork(const BlockHash &hash, uint32_t nBits,
diff --git a/src/pow/test/CMakeLists.txt b/src/pow/test/CMakeLists.txt
index a6aad13f9e713fc27dd9ac06524b864f5cb309fa..922c920680235522ab05d5ff06facd0c4f16a69d 100644
--- a/src/pow/test/CMakeLists.txt
+++ b/src/pow/test/CMakeLists.txt
@@ -12,7 +12,8 @@ add_boost_unit_tests_to_suite(pow test-pow
 	fixture.cpp
 
 	TESTS
-		eda_tests.cpp
+		daa_tests.cpp
+		dogecoin_daa_tests.cpp
 )
 
 target_link_libraries(test-pow server testutil)
diff --git a/src/pow/test/eda_tests.cpp b/src/pow/test/daa_tests.cpp
similarity index 53%
rename from src/pow/test/eda_tests.cpp
rename to src/pow/test/daa_tests.cpp
index cdd6a87982a4c3f23ea4268f223dc38e24b877b9..c5449664ddc08760a1e992b1158369770867f17d 100644
--- a/src/pow/test/eda_tests.cpp
+++ b/src/pow/test/daa_tests.cpp
@@ -2,7 +2,6 @@
 // Distributed under the MIT/X11 software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#include <pow/eda.h>
 #include <pow/pow.h>
 
 #include <chain.h>
@@ -16,108 +15,7 @@
 
 #include <boost/test/unit_test.hpp>
 
-BOOST_FIXTURE_TEST_SUITE(eda_tests, BasicTestingSetup)
-
-/* Test calculation of next difficulty target with no constraints applying */
-BOOST_AUTO_TEST_CASE(get_next_work) {
-    DummyConfig config(ChainTypeToString(ChainType::MAIN));
-
-    int64_t nLastRetargetTime = 1261130161; // Block #30240
-    CBlockIndex pindexLast;
-    pindexLast.nHeight = 32255;
-    pindexLast.nTime = 1262152739; // Block #32255
-    pindexLast.nBits = 0x1d00ffff;
-
-    // Here (and below): expected_nbits is calculated in
-    // CalculateNextWorkRequired(); redoing the calculation here would be just
-    // reimplementing the same code that is written in pow.cpp. Rather than
-    // copy that code, we just hardcode the expected result.
-    unsigned int expected_nbits = 0x1d00d86aU;
-    auto consensus_params = config.GetChainParams().GetConsensus();
-    BOOST_CHECK_EQUAL(CalculateNextWorkRequired(&pindexLast, nLastRetargetTime,
-                                                consensus_params),
-                      expected_nbits);
-    BOOST_CHECK(
-        PermittedDifficultyTransition(consensus_params, pindexLast.nHeight + 1,
-                                      pindexLast.nBits, expected_nbits));
-}
-
-/* Test the constraint on the upper bound for next work */
-BOOST_AUTO_TEST_CASE(get_next_work_pow_limit) {
-    DummyConfig config(ChainTypeToString(ChainType::MAIN));
-
-    int64_t nLastRetargetTime = 1231006505; // Block #0
-    CBlockIndex pindexLast;
-    pindexLast.nHeight = 2015;
-    pindexLast.nTime = 1233061996; // Block #2015
-    pindexLast.nBits = 0x1d00ffff;
-    unsigned int expected_nbits = 0x1d00ffffU;
-    auto consensus_params = config.GetChainParams().GetConsensus();
-
-    // TODO: Temporary fix for Dogecoin: Restore Bitcoin parameters
-    // Replace this test suite with Dogecoin DAA tests
-    consensus_params.powLimit = uint256S(
-        "00000000ffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
-    consensus_params.nPowTargetTimespan = 14 * 24 * 60 * 60;
-    consensus_params.nPowTargetSpacing = 10 * 60;
-
-    BOOST_CHECK_EQUAL(CalculateNextWorkRequired(&pindexLast, nLastRetargetTime,
-                                                consensus_params),
-                      expected_nbits);
-    BOOST_CHECK(
-        PermittedDifficultyTransition(consensus_params, pindexLast.nHeight + 1,
-                                      pindexLast.nBits, expected_nbits));
-}
-
-/* Test the constraint on the lower bound for actual time taken */
-BOOST_AUTO_TEST_CASE(get_next_work_lower_limit_actual) {
-    DummyConfig config(ChainTypeToString(ChainType::MAIN));
-
-    int64_t nLastRetargetTime = 1279008237; // Block #66528
-    CBlockIndex pindexLast;
-    pindexLast.nHeight = 68543;
-    pindexLast.nTime = 1279297671; // Block #68543
-    pindexLast.nBits = 0x1c05a3f4;
-    unsigned int expected_nbits = 0x1c0168fdU;
-    auto consensus_params = config.GetChainParams().GetConsensus();
-    BOOST_CHECK_EQUAL(CalculateNextWorkRequired(&pindexLast, nLastRetargetTime,
-                                                consensus_params),
-                      expected_nbits);
-    BOOST_CHECK(
-        PermittedDifficultyTransition(consensus_params, pindexLast.nHeight + 1,
-                                      pindexLast.nBits, expected_nbits));
-    // Test that reducing nbits further would not be a
-    // PermittedDifficultyTransition.
-    unsigned int invalid_nbits = expected_nbits - 1;
-    BOOST_CHECK(
-        !PermittedDifficultyTransition(consensus_params, pindexLast.nHeight + 1,
-                                       pindexLast.nBits, invalid_nbits));
-}
-
-/* Test the constraint on the upper bound for actual time taken */
-BOOST_AUTO_TEST_CASE(get_next_work_upper_limit_actual) {
-    DummyConfig config(ChainTypeToString(ChainType::MAIN));
-
-    int64_t nLastRetargetTime = 1263163443; // NOTE: Not an actual block time
-    CBlockIndex pindexLast;
-    pindexLast.nHeight = 46367;
-    pindexLast.nTime = 1269211443; // Block #46367
-    pindexLast.nBits = 0x1c387f6f;
-    unsigned int expected_nbits = 0x1d00e1fdU;
-    auto consensus_params = config.GetChainParams().GetConsensus();
-    BOOST_CHECK_EQUAL(CalculateNextWorkRequired(&pindexLast, nLastRetargetTime,
-                                                consensus_params),
-                      expected_nbits);
-    BOOST_CHECK(
-        PermittedDifficultyTransition(consensus_params, pindexLast.nHeight + 1,
-                                      pindexLast.nBits, expected_nbits));
-    // Test that increasing nbits further would not be a
-    // PermittedDifficultyTransition.
-    unsigned int invalid_nbits = expected_nbits + 1;
-    BOOST_CHECK(
-        !PermittedDifficultyTransition(consensus_params, pindexLast.nHeight + 1,
-                                       pindexLast.nBits, invalid_nbits));
-}
+BOOST_FIXTURE_TEST_SUITE(daa_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(CheckProofOfWork_test_negative_target) {
     const auto consensus =
@@ -180,7 +78,7 @@ BOOST_AUTO_TEST_CASE(GetBlockProofEquivalentTime_test) {
         blocks[i].nTime =
             1269211443 +
             i * config.GetChainParams().GetConsensus().nPowTargetSpacing;
-        blocks[i].nBits = 0x207fffff; /* target 0x7fffff000... */
+        blocks[i].nBits = 0x207fffff; // target 0x7fffff000...
         blocks[i].nChainWork =
             i ? blocks[i - 1].nChainWork + GetBlockProof(blocks[i])
               : arith_uint256(0);
@@ -214,7 +112,8 @@ BOOST_AUTO_TEST_CASE(retargeting_test) {
 
     std::vector<CBlockIndex> blocks(115);
 
-    const Consensus::Params &params = config.GetChainParams().GetConsensus();
+    const CChainParams &chainParams = config.GetChainParams();
+    const Consensus::Params &params = chainParams.GetConsensus();
     const arith_uint256 powLimit = UintToArith256(params.powLimit);
     arith_uint256 currentPow = powLimit >> 1;
     uint32_t initialBits = currentPow.GetCompact();
@@ -241,7 +140,7 @@ BOOST_AUTO_TEST_CASE(retargeting_test) {
     for (size_t i = 100; i < 110; i++) {
         blocks[i] = GetBlockIndex(&blocks[i - 1], 2 * 3600, initialBits);
         BOOST_CHECK_EQUAL(
-            GetNextEDAWorkRequired(&blocks[i], &blkHeaderDummy, params),
+            GetNextWorkRequired(&blocks[i], &blkHeaderDummy, chainParams),
             initialBits);
     }
 
@@ -249,7 +148,7 @@ BOOST_AUTO_TEST_CASE(retargeting_test) {
     blocks[110] = GetBlockIndex(&blocks[109], 2 * 3600, initialBits);
     currentPow.SetCompact(currentPow.GetCompact());
     BOOST_CHECK_EQUAL(
-        GetNextEDAWorkRequired(&blocks[110], &blkHeaderDummy, params),
+        GetNextWorkRequired(&blocks[110], &blkHeaderDummy, chainParams),
         currentPow.GetCompact());
 
     // As we continue with 2h blocks, difficulty remains unchanged.
@@ -257,7 +156,7 @@ BOOST_AUTO_TEST_CASE(retargeting_test) {
         GetBlockIndex(&blocks[110], 2 * 3600, currentPow.GetCompact());
     currentPow.SetCompact(currentPow.GetCompact());
     BOOST_CHECK_EQUAL(
-        GetNextEDAWorkRequired(&blocks[111], &blkHeaderDummy, params),
+        GetNextWorkRequired(&blocks[111], &blkHeaderDummy, chainParams),
         currentPow.GetCompact());
 
     // Difficulty remains unchanged.
@@ -265,7 +164,7 @@ BOOST_AUTO_TEST_CASE(retargeting_test) {
         GetBlockIndex(&blocks[111], 2 * 3600, currentPow.GetCompact());
     currentPow.SetCompact(currentPow.GetCompact());
     BOOST_CHECK_EQUAL(
-        GetNextEDAWorkRequired(&blocks[112], &blkHeaderDummy, params),
+        GetNextWorkRequired(&blocks[112], &blkHeaderDummy, chainParams),
         currentPow.GetCompact());
 
     // Difficulty remains unchanged, so we don't go below powLimit anyway.
@@ -274,14 +173,14 @@ BOOST_AUTO_TEST_CASE(retargeting_test) {
     currentPow.SetCompact(currentPow.GetCompact());
     BOOST_CHECK(powLimit.GetCompact() != currentPow.GetCompact());
     BOOST_CHECK_EQUAL(
-        GetNextEDAWorkRequired(&blocks[113], &blkHeaderDummy, params),
+        GetNextWorkRequired(&blocks[113], &blkHeaderDummy, chainParams),
         currentPow.GetCompact());
 
     // Difficulty remains unchanged, so we don't go below powLimit anyway.
     blocks[114] = GetBlockIndex(&blocks[113], 2 * 3600, powLimit.GetCompact());
     BOOST_CHECK(powLimit.GetCompact() != currentPow.GetCompact());
     BOOST_CHECK_EQUAL(
-        GetNextEDAWorkRequired(&blocks[114], &blkHeaderDummy, params),
+        GetNextWorkRequired(&blocks[114], &blkHeaderDummy, chainParams),
         powLimit.GetCompact());
 }
 
diff --git a/src/pow/test/dogecoin_daa_tests.cpp b/src/pow/test/dogecoin_daa_tests.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..0264c338497dd0afe93e29181c7c978ddc47da55
--- /dev/null
+++ b/src/pow/test/dogecoin_daa_tests.cpp
@@ -0,0 +1,163 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <pow/pow.h>
+
+#include <chain.h>
+#include <chainparams.h>
+#include <config.h>
+
+#include <test/util/random.h>
+#include <test/util/setup_common.h>
+
+#include <boost/test/unit_test.hpp>
+
+std::vector<CBlockIndex> MakeMockBlocks(size_t length, int32_t startHeight) {
+    std::vector<CBlockIndex> blocks(length);
+    for (size_t i = 0; i < length; i++) {
+        blocks[i].pprev = i ? &blocks[i - 1] : nullptr;
+        blocks[i].nHeight = startHeight + i;
+    }
+    return blocks;
+}
+
+BOOST_FIXTURE_TEST_SUITE(dogecoin_daa_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(test_first_daa) {
+    DummyConfig config(CBaseChainParams::MAIN);
+    std::vector<CBlockIndex> blocks = MakeMockBlocks(240, 0);
+
+    CBlockHeader header;
+    blocks[0].nTime = 1386325540; // Block # 0
+
+    // f9533416310fc4484cf43405a858b06afc9763ad401d267c1835d77e7d225a4e
+    CBlockIndex *pindexLast = &blocks[239];
+    BOOST_CHECK_EQUAL(pindexLast->nHeight, 239);
+    pindexLast->nTime = 1386475638; // Block #239
+    pindexLast->nBits = 0x1e0ffff0;
+
+    const uint32_t expected_nbits = 0x1e0fffff;
+    BOOST_CHECK_EQUAL(
+        GetNextWorkRequired(pindexLast, &header, config.GetChainParams()),
+        expected_nbits);
+    BOOST_CHECK(PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits));
+}
+
+BOOST_AUTO_TEST_CASE(get_next_work_pre_digishield) {
+    DummyConfig config(CBaseChainParams::MAIN);
+    std::vector<CBlockIndex> blocks = MakeMockBlocks(241, 9359);
+
+    CBlockHeader header;
+    blocks[0].nTime = 1386942008; // Block # 9359
+
+    CBlockIndex *pindexLast = &blocks[240];
+    BOOST_CHECK_EQUAL(pindexLast->nHeight, 9599);
+    pindexLast->nTime = 1386954113;
+    pindexLast->nBits = 0x1c1a1206;
+
+    const uint32_t expected_nbits = 0x1c15ea59;
+    BOOST_CHECK_EQUAL(
+        GetNextWorkRequired(pindexLast, &header, config.GetChainParams()),
+        expected_nbits);
+    BOOST_CHECK(PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits));
+}
+
+BOOST_AUTO_TEST_CASE(get_next_work_digishield) {
+    DummyConfig config(CBaseChainParams::MAIN);
+    std::vector<CBlockIndex> blocks = MakeMockBlocks(2, 144999);
+
+    CBlockHeader header;
+    blocks[0].nTime = 1395094427; // Block # 144999
+
+    // First hard-fork at 145,000, which applies to block 145,001 onwards
+    CBlockIndex *pindexLast = &blocks[1];
+    BOOST_CHECK_EQUAL(pindexLast->nHeight, 145000);
+    pindexLast->nTime = 1395094679;
+    pindexLast->nBits = 0x1b499dfd;
+
+    const uint32_t expected_nbits = 0x1b671062;
+    BOOST_CHECK_EQUAL(
+        GetNextWorkRequired(pindexLast, &header, config.GetChainParams()),
+        expected_nbits);
+    BOOST_CHECK(PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits));
+}
+
+BOOST_AUTO_TEST_CASE(get_next_work_digishield_modulated_upper) {
+    DummyConfig config(CBaseChainParams::MAIN);
+    std::vector<CBlockIndex> blocks = MakeMockBlocks(2, 145106);
+
+    CBlockHeader header;
+    blocks[0].nTime = 1395100835; // Block # 145,106
+
+    // Test the upper bound on modulated time using mainnet block #145,107
+    CBlockIndex *pindexLast = &blocks[1];
+    BOOST_CHECK_EQUAL(pindexLast->nHeight, 145107);
+    pindexLast->nTime = 1395101360;
+    pindexLast->nBits = 0x1b3439cd;
+
+    const uint32_t expected_nbits = 0x1b4e56b3;
+    BOOST_CHECK_EQUAL(
+        GetNextWorkRequired(pindexLast, &header, config.GetChainParams()),
+        expected_nbits);
+    BOOST_CHECK(PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits));
+    BOOST_CHECK(!PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits + 1));
+}
+
+BOOST_AUTO_TEST_CASE(get_next_work_digishield_modulated_lower) {
+    DummyConfig config(CBaseChainParams::MAIN);
+    std::vector<CBlockIndex> blocks = MakeMockBlocks(2, 149422);
+
+    CBlockHeader header;
+    blocks[0].nTime = 1395380517; // Block # 149,422
+
+    // Test the lower bound on modulated time using mainnet block #149,423
+    CBlockIndex *pindexLast = &blocks[1];
+    BOOST_CHECK_EQUAL(pindexLast->nHeight, 149423);
+    pindexLast->nTime = 1395380447;
+    pindexLast->nBits = 0x1b446f21;
+    const uint32_t expected_nbits = 0x1b335358;
+    BOOST_CHECK_EQUAL(
+        GetNextWorkRequired(pindexLast, &header, config.GetChainParams()),
+        expected_nbits);
+    BOOST_CHECK(PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits));
+    BOOST_CHECK(!PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits - 1));
+}
+
+BOOST_AUTO_TEST_CASE(get_next_work_digishield_rounding) {
+    DummyConfig config(CBaseChainParams::MAIN);
+    std::vector<CBlockIndex> blocks = MakeMockBlocks(2, 145000);
+
+    CBlockHeader header;
+    blocks[0].nTime = 1395094679;
+
+    // Test case for correct rounding of modulated time - this depends on
+    // handling of integer division, and is not obvious from the code
+    CBlockIndex *pindexLast = &blocks[1];
+    BOOST_CHECK_EQUAL(pindexLast->nHeight, 145001);
+    pindexLast->nTime = 1395094727;
+    pindexLast->nBits = 0x1b671062;
+    const uint32_t expected_nbits = 0x1b6558a4;
+    BOOST_CHECK_EQUAL(
+        GetNextWorkRequired(pindexLast, &header, config.GetChainParams()),
+        expected_nbits);
+    BOOST_CHECK(PermittedDifficultyTransition(
+        config.GetChainParams().GetConsensus(), pindexLast->nHeight + 1,
+        pindexLast->nBits, expected_nbits));
+}
+
+BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index fa7bdad677a543749a1682539f0bc8e1bc39c7e7..93eec38094ce4a1fd56f67cf6636db6e13d39db0 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -73,9 +73,11 @@ static UniValue GetNetworkHashPS(int lookup, int height,
 
     // If lookup is -1, then use blocks since last difficulty change.
     if (lookup <= 0) {
-        lookup = pb->nHeight %
-                     Params().GetConsensus().DifficultyAdjustmentInterval() +
-                 1;
+        const Consensus::Params &params = Params().GetConsensus();
+        const Consensus::DaaParams daaParams =
+            params.DaaParamsAtHeight(pb->nHeight);
+        lookup =
+            pb->nHeight % params.DifficultyAdjustmentInterval(daaParams) + 1;
     }
 
     // If lookup is larger than chain, then set it to chain length.
diff --git a/src/test/fuzz/pow.cpp b/src/test/fuzz/pow.cpp
index 6a2b3505f5b880d2e0ef10dcc0bd63190f145650..cc99fb6a9364cc6b9bf6cb72ca37abf0675d9b5a 100644
--- a/src/test/fuzz/pow.cpp
+++ b/src/test/fuzz/pow.cpp
@@ -83,7 +83,8 @@ FUZZ_TARGET_INIT(pow, initialize_pow) {
             (void)GetBlockProof(current_block);
             if (current_block.nHeight != std::numeric_limits<int>::max() &&
                 current_block.nHeight -
-                        (consensus_params.DifficultyAdjustmentInterval() - 1) >=
+                        (consensus_params.DifficultyAdjustmentInterval({}) -
+                         1) >=
                     0) {
                 (void)GetNextWorkRequired(&current_block, &(*block_header),
                                           chainparams);
